File: with_arma/Rcpp/perfRcpp/R/perf.R
assert = function(bool) {
    if (!bool) stop('Assertion failed')
}

perf <- function(){
    library(microbenchmark)
    filename <- "Rcpptimes.csv"
    conv_nanosecond_to_second <- 1e-9
    ans_pisum <- 1.644834071848065
    ans_fib <- 832040
    # should be ans_pisum <- pi^2 / 6
    # 1.6449340668482264061

    #====================================================#
    # R function calling Rcpp function calling cpp function
    assert(abs(pisumR()-ans_pisum) < 1e-12);
    # using nanoseconds, because seems to be buggy askign for seconds
    # unit = nanoseconds
    # options(microbenchmark.unit="ns")
    res <- microbenchmark(pisumR(), times=21L, unit="ns")
    tmed_pisum <- median(res$time)

    # now convert to seconds
    tmed_pisum <- tmed_pisum * conv_nanosecond_to_second
    cat("pisumR: ", tmed_pisum, " seconds \n", sep="")
    #====================================================#

    #====================================================#
    # Rcpp function, calling cpp function
    assert(abs(pisumRcpp()-ans_pisum) < 1e-12);
    # using nanoseconds, because seems to be buggy askign for seconds
    # unit = nanoseconds
    # options(microbenchmark.unit="ns")
    res <- microbenchmark(pisumRcpp(), times=21L, unit="ns")
    tmed_pisum2 <- median(res$time)

    # now convert to seconds
    tmed_pisum2 <- tmed_pisum2 * conv_nanosecond_to_second
    cat("pisumRcpp: ", tmed_pisum2, " seconds \n", sep="")
    #====================================================#

    #====================================================#
    # R function calling Rcpp function calling cpp function
    assert( fibR(30) == ans_fib )
    res <- microbenchmark(fibR(30), times=21L, unit="ns")
    tmed_fib <- median(res$time)

    # now convert to seconds
    tmed_fib <- tmed_fib * conv_nanosecond_to_second
    cat("fibR (calling Rcpp calling cpp): ", tmed_fib, " seconds \n", sep="")
    #====================================================#


    #====================================================#
    # Rcpp function calling cpp function
    assert( fibRcpp(30) == ans_fib )
    res <- microbenchmark(fibRcpp(30), times=21L, unit="ns")
    tmed_fib2 <- median(res$time)

    # now convert to seconds
    tmed_fib2 <- tmed_fib2 * conv_nanosecond_to_second
    cat("fibRcpp (calling cpp): ", tmed_fib2, " seconds \n", sep="")
    #====================================================#

    #====================================================#
    # Rcpp function for randmatstats will not work on OSX, 
    # because having trouble with RppArmadillo installation
    tmed_randmatstats <- 0
    cat("randmatstatsRcpp (not working): ", tmed_randmatstats, " seconds \n", sep="")
    #====================================================#


    tmed <- c(tmed_pisum, tmed_fib, tmed_randmatstats)
    algoNames <- c("pisum", "fib", "randmatstats")
    df <- data.frame(Rcpp=tmed)
    row.names(df) <- algoNames

    # write to table
    write.table(file=filename, x=df, sep=",", row.names=T, col.names=T, quote=F)
    cat("Writing to: ", filename, "\n", sep="")
}

File: no_arma/cython/perf.py
from perf_module import cy_pisum
from perf_module import cy_fib
import time


if __name__=="__main__":
    filename = "cythontimes.txt"

    trials= 21
    assert abs(cy_pisum()-1.6448340718480652) < 1e-9
    time_vec = [0.0] * trials
    for i in range(trials):
        t = time.time()
        pi = cy_pisum()
        t = time.time()-t
        time_vec[i] = t
    # get median time
    med_index = int(trials / 2)
    time_vec.sort()
    tmed_pisum = time_vec[med_index]
    print("Cython, pisum: ", tmed_pisum, " seconds")
    print(pi)


    assert cy_fib(30) == 832040
    time_vec = [0.0] * trials
    for i in range(trials):
        t = time.time()
        fibnum = cy_fib(30)
        t = time.time()-t
        time_vec[i] = t
    med_index = int(trials / 2)
    time_vec.sort()
    tmed_fib = time_vec[med_index]
    print("Cython, fib: ", tmed_fib, " seconds")
    print(fibnum)


    # write results to file
    f=open(filename,"w+")
    f.write("Cython\n")
    f.write("pisum," + str(tmed_pisum) + "\n")
    f.write("fib," + str(tmed_fib) + "\n")
    f.close()

    print("Writing results to: ", filename, "\n")
